### PR TITLE
Finding the Center

### PR DESCRIPTION
**Pull request by prajy**
Added the  code on how to find it with bounding rectangles and moments (the current one is the aspect ratio code). I am not sure how to make the  Cx=M10/M00  and Cy=M01/M00 look fancy in bold and formula like.

### CODE REVIEW

**Comment by TuringsEgo** [comment]
Would you want to talk about the advantages of using moments vs finding the center of a bounding box? (Numerical stability). If not, I can throw in a case study of why extracting the center from anything other than the moment can be dangerous.

**Comment by prajy** [comment]
I'm actually not sure what th big differences are but I can look into it. I just used the existing description. You think having both the differences and a case study is a good idea? And there probably should be a bit more information on what exactly a moment is and that it factors in Gravity and stuff when calculating the center.

**Comment by TuringsEgo** [comment]
There really isn't much to moments beyond the equation. Either you understand it or you don't. That being said, morphological operations goes into quite a bit of set theory math. 

I think a case study of some sort would be good. Show how a noisy contour can really mess up the center calculation when you extract from bounding rectangle. 

@Twinters007 what do you think?

**Comment by Twinters007** [comment]
idk if we want to get rid of the moment code. It makes more sense to add on to it with bounding box

**Comment by prajy** [comment]
I just added a missing parenthesis in the C++ tab for contour area. Also, The contour area says if(contourArea < min_area || contourArea > maxArea). However shouldnt it be contourArea is greater than the min area and contour area less than max area? So this if(contourArea > min_area || contourArea < maxArea).

**Comment by Twinters007** [comment]
No because if it's greater than the max or less than min, ignore it (`continue`)

**Comment by prajy** [comment]
oops

**Comment by prajy** [comment]
lol can you just delete that commit right there

**Comment by prajy** [comment]
if you cant ill just reset the commit (or try to lol)

**Comment by prajy** [comment]
fixed

**Comment by prajy** [comment]
but could you do all the filtering in one for loop like this:

        for(int i = 0; i< contours.size(); i++ ) // iterate through each contour. 
        {
         Rect boundRect = boundingRect(contours[i]);
         float contourArea = contourArea(contours[i]);  
         float aspectRatio = (float)boundRect.width/boundRect.height 
         float solidityRatio = contourArea(contours[i])/(boundRect.width*boundRect.height); 
         if (contourArea > maxArea || contourArea < minArea)
          {
              continue;
         }
         if (aspectRatio > maxAspectRatio || aspectRatio < minAspectRatio)
          {
              continue;
         }
         if (solidityRatio > maxSolidityRatio || solidityRatio < minASolidityRatio)
          {
              continue;
         }
         }

**Comment by Twinters007** [comment]
Yeah you could if you were using all 3 of those factors. But since we only show how to filter by area...eh

**Comment by Twinters007** [comment]
We should define what centroid means

**Comment by prajy** [comment]
Yeah.

**Comment by Twinters007** [comment]
Define what centroid means and I'll merge

**Comment by prajy** [comment]
added def of centroid. I tried to add in subscripts with sphinx notation for the moment centroid formula but looks like i did something wrong.

**Comment by Twinters007** [comment]
:math: \`C_x = M_{10} / M_{00}` and :math:`C_y =M_{01} /M_{00}\`.

File: vision/contours.rst [hunk 1]
```diff
@@ -168,26 +168,50 @@ The last test we will cover is solidity. That is, the ratio between the contour
...
 
    .. code-tab:: py
 
          x,y,w,h = cv2.boundingRect(contour)
-         ratio = cv2.contourArea(contour)/(w*h)
+	 double centerY = boundRect.y + (boundRect.h / 2)
+         
+
+
+While you could simply apply a bounded rectangle and then find the center of that, there is a more precise way: N-th order moments. Mathematically, a moment is defined as :math:`\mu _{n}=\int _{-\infty }^{\infty }(x-c)^{n}\,f(x)\,dx`. For a 2D continuous function f(x,y) the moment of order (p + q) is defined as :math:`M_{{pq}}=\int \limits _{{-\infty }}^{{\infty }}\int \limits _{{-\infty }}^{{\infty }}x^{p}y^{q}f(x,y)\,dx\,dy`. The area of a contour is the zeroth moment, and moments can be used to find the centroid of a contour. The centroid is the center of mass of an object. The centroid point is also the center of gravity. Using moments, the centroid, C is defined as :math: `C_x = M_{10} / M_{00}and :math:C_y =M_{01} /M_{00}`.
```
Comment (Twinters007) [inline vision/contours.rst]: You need another backtick after `{00}`. (Hint: test stuff)

File: vision/contours.rst [hunk 2]
```diff
@@ -168,26 +168,50 @@ The last test we will cover is solidity. That is, the ratio between the contour
...
 
    .. code-tab:: py
 
          x,y,w,h = cv2.boundingRect(contour)
-         ratio = cv2.contourArea(contour)/(w*h)
+	 double centerY = boundRect.y + (boundRect.h / 2)
+         
+
+
+While you could simply apply a bounded rectangle and then find the center of that, there is a more precise way: N-th order moments. Mathematically, a moment is defined as :math:`\mu _{n}=\int _{-\infty }^{\infty }(x-c)^{n}\,f(x)\,dx`. For a 2D continuous function f(x,y) the moment of order (p + q) is defined as :math:`M_{{pq}}=\int \limits _{{-\infty }}^{{\infty }}\int \limits _{{-\infty }}^{{\infty }}x^{p}y^{q}f(x,y)\,dx\,dy`. The area of a contour is the zeroth moment, and moments can be used to find the centroid of a contour. The centroid is the center of mass of an object. The centroid point is also the center of gravity. Using moments, the centroid, C is defined as C:sub:`x`=M:sub:`10`/M:sub:`00` and C:sub:`y`=M:sub:`01`/M:sub:`00`.
```
Comment (Twinters007) [inline vision/contours.rst]: wait now its back to the old thing

**Comment by prajy** [comment]
i fixed it before looking at this lol

**Comment by prajy** [comment]
i did check right now

**Pull request merged**